python_ta/cfg/graph.py

```diff
@@ -245,3 +341,86 @@ def __init__(
         self.condition = condition
         self.source.successors.append(self)
         self.target.predecessors.append(self)
+        self.z3_constraints = {}
+
+
+class Z3Environment:
+    """Z3 Environment stores the Z3 variables and constraints in the current CFG path
+
+    variable_unassigned:
+        A dictionary mapping each variable in the current environment to a boolean indicating
+        whether it has been reassigned (False) or remains unassigned (True).
+
+    variable_type:
+        A dictionary mapping each variable in the current environment to its data type.
+
+    constraints:
+        A list of Z3 constraints in the current environment.
+    """
+
+    variable_unassigned: Dict[str, bool]
+    variable_type: Dict[str, str]
+    constraints: List[ExprRef]
+
+    def __init__(self, variables: Dict[str, ExprRef], constraints: List[ExprRef]) -> None:
+        """Initialize the environment with function parameters and preconditions"""
+        self.variable_unassigned = {var: True for var in variables.keys()}
+        self.variable_type = variables
+        self.constraints = constraints.copy()
+
+    def assign(self, name: str) -> None:
+        """Handle a variable assignment statement"""
+        if name in self.variable_unassigned:
+            self.variable_unassigned[name] = False
+
+    def update_constraints(self) -> List[ExprRef]:
+        """
+        Returns all z3 constraints in the environments
+        Removes constraints with reassigned variables
+        """
+        updated_constraints = []
+        for constraint in self.constraints:
+            # discard expressions with reassigned variables
+            variables = _get_vars(constraint)
+            reassigned = any(
+                not self.variable_unassigned.get(variable, False) for variable in variables
+            )
+            if not reassigned:
+                updated_constraints.append(constraint)
+
+        self.constraints = updated_constraints
+        return updated_constraints.copy()
+
+    def add_constraint(self, constraint: ExprRef) -> None:
+        """
+        Add a new z3 constraint to environment
+        """
+        self.constraints.append(constraint)
+
+    def parse_constraint(self, node: NodeNG) -> Optional[ExprRef]:
+        """
+        Parse an Astroid node to a Z3 constraint
+        Return the resulting expression
+        """
+        ew = ExprWrapper(node, self.variable_type)
+        try:
+            return ew.reduce()
+        except (Z3Exception, Z3ParseException):
+            return None
+
+
+def _get_vars(expr: ExprRef) -> Set[str]:
+    """
+    Retrieve all z3 variables from a z3 expression
+    """
+    variables = set()
+
+    def traverse(e):
```

Comment:
Similar to my previous comment, look back at the feedback in my previous review and apply it here as well.